File: BlogModule.Application/CommandHandlers/BlogPost/CreateBlogPost/CreateBlogPostCommandHandler.cs
using BlogModule.Application.Contract.BlogPosts.CreateBlogPost;
using BlogModule.Domain.Models;
using BlogModule.Domain.RepositoryInterfaces.BlogPost;
using Results.Application;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.CommandHandlers;

namespace BlogModule.Application.CommandHandlers.BlogPost.CreateBlogPost;

public class CreateBlogPostCommandHandler : ICommandHandler<CreateBlogPostCommand, HandlerResult<CreateBlogPostResponse, IErrorResult>>
{
    private readonly IBlogPostCommandRepository _commandRepository;
    private readonly IBlogPostQueryRepository _queryRepository;
    
    public CreateBlogPostCommandHandler(
        IBlogPostCommandRepository repository, 
        IBlogPostQueryRepository queryRepository)
    {
        _commandRepository = repository;
        _queryRepository = queryRepository;
    }
    
    public async Task<HandlerResult<CreateBlogPostResponse, IErrorResult>> Handle(CreateBlogPostCommand request, CancellationToken cancellationToken)
    {
        var blogPost = await _queryRepository.GetByUrlAsync(request.UrlHandle, cancellationToken);
        if (blogPost is not null) return new EntityAlreadyExistsErrorResult() {
            Title = "Cannot create blog post",
            Message = $"Blog post with url handle {request.UrlHandle} already exists in the database."
        };
        
        var blogPostToAdd = new BlogPostModel
        {
            Heading = request.Heading,
            PageTitle = request.PageTitle,
            Content = request.Content,
            ShortDescription = request.ShortDescription,
            FeaturedImageUrl = request.FeaturedImageUrl,
            UrlHandle = request.UrlHandle,
            PublishedDate = request.PublishedDate,
            Author = request.Author,
            Visible = request.Visible,
            Tags = request.Tags,
            Likes = blogPost.Likes,
            Comments = blogPost.Comments
        };

        var addedBlogPost = await _commandRepository.AddAsync(blogPostToAdd, cancellationToken);

        return new CreateBlogPostResponse() {
            Id = addedBlogPost.Id,
            Title = "Blog post created",
            Message = $"Blog post '{addedBlogPost.Heading}' was created in the database."
        };
    }
}

File: BlogApi/Application/command_handlers/blog_post/create_blog_post/create_blog_post_command_handler.go
package commands

import (
	contract "blog-api/Application.contract/BlogPosts/CreateBlogPost"
	models "blog-api/Domain/models/domestic"
	repository_interfaces "blog-api/Domain/repository_interfaces/blog_post_repository"
	"blog-api/Domain/responses"
	data_fetcher "blog-api/Domain/service_interfaces"
	"context"
	"time"
)

type CreateBlogPostCommandHandler struct {
	blogPostCommandRepository repository_interfaces.BlogPostCommandRepository
	dataFetcher               data_fetcher.DataFetcher
}

func NewCreateBlogPostCommandHandler(
	blogPostCommandRepository repository_interfaces.BlogPostCommandRepository,
	dataFetcher data_fetcher.DataFetcher,
) *CreateBlogPostCommandHandler {
	return &CreateBlogPostCommandHandler{
		blogPostCommandRepository: blogPostCommandRepository,
		dataFetcher:               dataFetcher,
	}
}

func (h *CreateBlogPostCommandHandler) Handle(ctx context.Context, command *CreateBlogPostCommand) (*contract.CreateBlogPostResponse, error) {
	userInfo, err := h.dataFetcher.GetUserInternalInfo(command.JwtToken)
	if err != nil {
		response := responses.NewResponse[contract.CreateBlogPostResponse](401, "Unauthorized")
		return &response, nil
	}

	now := time.Now()
	blogPost := &models.BlogPostRequestModel{
		Heading:          command.Heading,
		PageTitle:        command.PageTitle,
		Content:          command.Content,
		ShortDescription: command.ShortDescription,
		FeaturedImageUrl: command.FeaturedImageUrl,
		UrlHandle:        command.UrlHandle,
		AuthorId:         userInfo.ID,
		AuthorName:       userInfo.Name,
		Visible:          command.Visible,
		CreatedAt:        now.Format(time.RFC3339),
		Tags:             command.Tags,
	}

	result, err := h.blogPostCommandRepository.CreateBlogPost(ctx, blogPost)
	if err != nil {
		response := responses.NewResponse[contract.CreateBlogPostResponse](500, "Failed to create blog post")
		return &response, nil
	}

	return &contract.CreateBlogPostResponse{
		BaseResponse: responses.NewBaseResponse(200, "Blog post created successfully"),
		Id:           *result,
	}, nil
}
